File: fractal_sequences/explorer_session.py
"""Portable session data for the interactive sequence explorer."""

from __future__ import annotations

import csv
import json
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any, Iterable

from .config import (
    AnalysisConfig,
    AxisRangeConfig,
    DiscretizationConfig,
    ExplorerConfig,
    RecurrenceConfig,
)
from .models import Attractor, Number, SequenceAnalysisResult, SequenceTerm
from .transformations import DiscreteAnalysisResult, DiscretePoint


SESSION_SCHEMA = "fractal_sequences.visual_session"
SESSION_VERSION = 1


@dataclass(frozen=True)
class ExplorationCase:
    """A named recurrence kept in the explorer's multi-case collection."""

    identifier: str
    label: str
    recurrence: RecurrenceConfig
    source: str = "manual"
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class ImportedSession:
    """Validated state restored from an exported JSON session."""

    config: ExplorerConfig
    cases: list[ExplorationCase]
    active_case_id: str | None = None
    table_mode: str = "todo"
    map_limits: dict[str, tuple[float, float, float, float]] = field(
        default_factory=dict
    )
    music: dict[str, str] = field(default_factory=dict)


def config_to_data(config: ExplorerConfig) -> dict[str, Any]:
    """Encode a pipeline configuration using JSON-native values."""

    return {
        "recurrence": {
            "name": config.recurrence.name,
            "parameters": dict(config.recurrence.parameters),
        },
        "analysis": asdict(config.analysis),
        "discretization": {
            "coordinate_system": config.discretization.coordinate_system,
            "ranges": {
                axis: asdict(axis_range)
                for axis, axis_range in config.discretization.ranges.items()
            },
        },
    }


def config_from_data(data: dict[str, Any]) -> ExplorerConfig:
    """Decode and validate a JSON-native pipeline configuration."""

    try:
        recurrence_data = data["recurrence"]
        analysis_data = data["analysis"]
        discretization_data = data["discretization"]
        ranges_data = discretization_data["ranges"]
    except (KeyError, TypeError) as exc:
        raise ValueError("Configuracion de sesion incompleta") from exc

    if not isinstance(recurrence_data, dict) or not isinstance(analysis_data, dict):
        raise ValueError("Configuracion de sesion invalida")
    if not isinstance(discretization_data, dict) or not isinstance(ranges_data, dict):
        raise ValueError("Discretizacion de sesion invalida")

    return ExplorerConfig(
        recurrence=RecurrenceConfig(
            name=str(recurrence_data["name"]),
            parameters={
                str(key): float(value)
                for key, value in dict(recurrence_data.get("parameters", {})).items()
            },
        ),
        analysis=AnalysisConfig(**analysis_data),
        discretization=DiscretizationConfig(
            coordinate_system=str(discretization_data["coordinate_system"]),
            ranges={
                str(axis): AxisRangeConfig(**axis_data)
                for axis, axis_data in ranges_data.items()
            },
        ),
    )


def case_to_data(case: ExplorationCase) -> dict[str, Any]:
    """Encode one stored case."""

    return {
        "identifier": case.identifier,
        "label": case.label,
        "recurrence": {
            "name": case.recurrence.name,
            "parameters": dict(case.recurrence.parameters),
        },
        "source": case.source,
        "metadata": case.metadata,
    }


def case_from_data(data: dict[str, Any]) -> ExplorationCase:
    """Decode one stored case."""

    recurrence_data = data.get("recurrence")
    if not isinstance(recurrence_data, dict):
        raise ValueError("Caso de sesion sin recurrencia")

    identifier = str(data.get("identifier", "")).strip()
    if not identifier:
        raise ValueError("Caso de sesion sin identificador")

    metadata = data.get("metadata", {})
    if not isinstance(metadata, dict):
        raise ValueError("Metadatos de caso invalidos")

    return ExplorationCase(
        identifier=identifier,
        label=str(data.get("label", identifier)),
        recurrence=RecurrenceConfig(
            name=str(recurrence_data["name"]),
            parameters={
                str(key): float(value)
                for key, value in dict(recurrence_data.get("parameters", {})).items()
            },
        ),
        source=str(data.get("source", "manual")),
        metadata=metadata,
    )


def result_to_data(result: DiscreteAnalysisResult) -> dict[str, Any]:
    """Encode all calculated values, including their discrete projections."""

    sequence = result.sequence_result
    return {
        "config": config_to_data(result.config) if isinstance(result.config, ExplorerConfig) else None,
        "behavior": sequence.behavior,
        "reason": sequence.reason,
        "diverged": sequence.diverged,
        "terms": [term_to_data(term) for term in sequence.terms],
        "attractors": [attractor_to_data(attractor) for attractor in sequence.attractors],
        "discrete_terms": [
            {
                "index": item.term.index,
                "point": point_to_data(item.point),
            }
            for item in result.terms
        ],
        "discrete_attractors": [
            {
                "kind": item.attractor.kind,
                "period": item.attractor.period,
                "points": [point_to_data(point) for point in item.points],
            }
            for item in result.attractors
        ],
        "coordinate_system": result.transformer.coordinate_system.name,
        "grid": {
            axis: {
                "minimum": subdivision.range.minimum,
                "maximum": subdivision.range.maximum,
                "steps": subdivision.steps,
            }
            for axis, subdivision in result.transformer.grid.subdivisions.items()
        },
    }


def build_session_data(
    config: ExplorerConfig,
    cases: Iterable[ExplorationCase],
    results: Iterable[tuple[ExplorationCase, DiscreteAnalysisResult]],
    *,
    active_case_id: str | None,
    table_mode: str,
    map_limits: dict[str, tuple[float, float, float, float]],
    music: dict[str, str] | None = None,
) -> dict[str, Any]:
    """Build a complete JSON document without writing it to disk."""

    return {
        "schema": SESSION_SCHEMA,
        "version": SESSION_VERSION,
        "view": {
            "config": config_to_data(config),
            "active_case_id": active_case_id,
            "table_mode": table_mode,
            "map_limits": {
                key: list(value)
                for key, value in map_limits.items()
            },
            "music": dict(music or {}),
        },
        "cases": [case_to_data(case) for case in cases],
        "results": [
            {
                "case_id": case.identifier,
                "case_label": case.label,
                "result": result_to_data(result),
            }
            for case, result in results
        ],
    }


def parse_session_data(data: dict[str, Any]) -> ImportedSession:
    """Validate an exported JSON document and restore its editable state."""

    if data.get("schema") != SESSION_SCHEMA:
        raise ValueError("El archivo no es una sesion de fractal_sequences")
    if data.get("version") != SESSION_VERSION:
        raise ValueError("Version de sesion no compatible")

    view = data.get("view")
    case_data = data.get("cases")
    if not isinstance(view, dict) or not isinstance(case_data, list):
        raise ValueError("Sesion incompleta")

    config = config_from_data(view.get("config", {}))
    cases = [case_from_data(item) for item in case_data]
    identifiers = [case.identifier for case in cases]
    if len(set(identifiers)) != len(identifiers):
        raise ValueError("La sesion contiene identificadores duplicados")

    raw_limits = view.get("map_limits", {})
    map_limits: dict[str, tuple[float, float, float, float]] = {}
    if isinstance(raw_limits, dict):
        for key, values in raw_limits.items():
            if isinstance(values, list) and len(values) == 4:
                map_limits[str(key)] = tuple(float(value) for value in values)

    active_case_id = view.get("active_case_id")
    if active_case_id is not None:
        active_case_id = str(active_case_id)
        if active_case_id not in identifiers:
            active_case_id = None

    raw_music = view.get("music", {})
    music = (
        {str(key): str(value) for key, value in raw_music.items()}
        if isinstance(raw_music, dict)
        else {}
    )

    return ImportedSession(
        config=config,
        cases=cases,
        active_case_id=active_case_id,
        table_mode=str(view.get("table_mode", "todo")),
        map_limits=map_limits,
        music=music,
    )


def write_session(path: str | Path, data: dict[str, Any]) -> None:
    """Write a session document using UTF-8 JSON."""

    Path(path).write_text(
        json.dumps(data, indent=2, ensure_ascii=False),
        encoding="utf-8",
    )


def read_session(path: str | Path) -> ImportedSession:
    """Read and validate a session document."""

    try:
        data = json.loads(Path(path).read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise ValueError("El archivo JSON no es valido") from exc
    if not isinstance(data, dict):
        raise ValueError("La raiz de la sesion debe ser un objeto JSON")
    return parse_session_data(data)


def write_results_csv(
    path: str | Path,
    results: Iterable[tuple[ExplorationCase, DiscreteAnalysisResult]],
) -> None:
    """Export sequence and attractor values for all cases to one CSV file."""

    columns = [
        "case_id",
        "case_label",
        "source",
        "row_type",
        "attractor_index",
        "term_index",
        "value_real",
        "value_imag",
        "coordinates",
        "indices",
        "discrete_coordinates",
        "cell_id",
        "inside_domain",
        "attractor_kind",
        "attractor_period",
    ]
    with Path(path).open("w", newline="", encoding="utf-8") as output:
        writer = csv.DictWriter(output, fieldnames=columns)
        writer.writeheader()
        for case, result in results:
            for item in result.terms:
                writer.writerow(
                    _csv_row(
                        case,
                        "sequence",
                        item.point,
                        term_index=item.term.index,
                    )
                )
            for attractor_index, attractor in enumerate(result.attractors, start=1):
                for point in attractor.points:
                    writer.writerow(
                        _csv_row(
                            case,
                            "attractor",
                            point,
                            attractor_index=attractor_index,
                            attractor_kind=attractor.attractor.kind,
                            attractor_period=attractor.attractor.period,
                        )
                    )


def term_to_data(term: SequenceTerm) -> dict[str, Any]:
    return {"index": term.index, "value": number_to_data(term.value)}


def attractor_to_data(attractor: Attractor) -> dict[str, Any]:
    return {
        "values": [number_to_data(value) for value in attractor.values],
        "kind": attractor.kind,
        "period": attractor.period,
        "tolerance": attractor.tolerance,
        "start_index": attractor.start_index,
    }


def point_to_data(point: DiscretePoint) -> dict[str, Any]:
    return {
        "value": number_to_data(point.value),
        "coordinates": point.coordinates,
        "indices": point.indices,
        "discrete_coordinates": point.discrete_coordinates,
        "cell_id": point.cell_id,
        "inside_domain": point.inside_domain,
    }


def number_to_data(value: Number) -> dict[str, float]:
    """Represent a real or complex number without losing its imaginary part."""

    complex_value = complex(value)
    return {"real": complex_value.real, "imag": complex_value.imag}


def _csv_row(
    case: ExplorationCase,
    row_type: str,
    point: DiscretePoint,
    *,
    attractor_index: int | None = None,
    term_index: int | None = None,
    attractor_kind: str | None = None,
    attractor_period: int | None = None,
) -> dict[str, Any]:
    value = complex(point.value)
    return {
        "case_id": case.identifier,
        "case_label": case.label,
        "source": case.source,
        "row_type": row_type,
        "attractor_index": attractor_index,
        "term_index": term_index,
        "value_real": value.real,
        "value_imag": value.imag,
        "coordinates": json.dumps(point.coordinates, ensure_ascii=False),
        "indices": json.dumps(point.indices, ensure_ascii=False),
        "discrete_coordinates": json.dumps(
            point.discrete_coordinates,
            ensure_ascii=False,
        ),
        "cell_id": point.cell_id,
        "inside_domain": point.inside_domain,
        "attractor_kind": attractor_kind,
        "attractor_period": attractor_period,
    }

File: src/explorer_session.rs
use std::collections::BTreeMap;
use std::path::Path;

use anyhow::{Result, bail};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

use crate::config::{AnalysisConfig, DiscretizationConfig, ExplorerConfig, RecurrenceConfig};
use crate::models::{Attractor, SequenceTerm};
use crate::transformations::{DiscreteAnalysisResult, DiscretePoint};

pub const SESSION_SCHEMA: &str = "fractal_sequences.visual_session";
pub const SESSION_VERSION: u32 = 1;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExplorationCase {
    pub identifier: String,
    pub label: String,
    pub recurrence: RecurrenceConfig,
    pub source: String,
    pub metadata: BTreeMap<String, String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ImportedSession {
    pub config: ExplorerConfig,
    pub cases: Vec<ExplorationCase>,
    pub active_case_id: Option<String>,
    pub table_mode: String,
    pub map_limits: BTreeMap<String, (f64, f64, f64, f64)>,
    pub music: BTreeMap<String, String>,
}

pub fn config_to_data(config: &ExplorerConfig) -> Value {
    json!({
        "recurrence": {
            "name": config.recurrence.name,
            "parameters": config.recurrence.parameters,
        },
        "analysis": config.analysis,
        "discretization": {
            "coordinate_system": config.discretization.coordinate_system,
            "ranges": config.discretization.ranges,
        }
    })
}

pub fn config_from_data(data: &Value) -> Result<ExplorerConfig> {
    let object = data
        .as_object()
        .ok_or_else(|| anyhow::anyhow!("Configuracion de sesion invalida"))?;
    let recurrence_data = object
        .get("recurrence")
        .ok_or_else(|| anyhow::anyhow!("Configuracion de sesion incompleta"))?;
    let analysis_data = object
        .get("analysis")
        .ok_or_else(|| anyhow::anyhow!("Configuracion de sesion incompleta"))?;
    let discretization_data = object
        .get("discretization")
        .ok_or_else(|| anyhow::anyhow!("Configuracion de sesion incompleta"))?;

    let recurrence = serde_json::from_value::<RecurrenceConfig>(recurrence_data.clone())?;
    let analysis = serde_json::from_value::<AnalysisConfig>(analysis_data.clone())?;
    let discretization = serde_json::from_value::<DiscretizationConfig>(discretization_data.clone())?;
    Ok(ExplorerConfig {
        recurrence,
        analysis,
        discretization,
    })
}

pub fn case_to_data(case: &ExplorationCase) -> Value {
    json!({
        "identifier": case.identifier,
        "label": case.label,
        "recurrence": case.recurrence,
        "source": case.source,
        "metadata": case.metadata,
    })
}

pub fn case_from_data(data: &Value) -> Result<ExplorationCase> {
    let case = serde_json::from_value::<ExplorationCase>(data.clone())?;
    if case.identifier.trim().is_empty() {
        bail!("Caso de sesion sin identificador");
    }
    Ok(case)
}

pub fn result_to_data(result: &DiscreteAnalysisResult) -> Value {
    json!({
        "config": result.config.as_ref().map(config_to_data),
        "behavior": result.sequence_result.behavior,
        "reason": result.sequence_result.reason,
        "diverged": result.sequence_result.diverged,
        "terms": result.sequence_result.terms.iter().map(term_to_data).collect::<Vec<_>>(),
        "attractors": result.sequence_result.attractors.iter().map(attractor_to_data).collect::<Vec<_>>(),
        "discrete_terms": result.terms.iter().map(|item| {
            json!({
                "index": item.term.index,
                "point": point_to_data(&item.point),
            })
        }).collect::<Vec<_>>(),
        "discrete_attractors": result.attractors.iter().map(|item| {
            json!({
                "kind": item.attractor.kind,
                "period": item.attractor.period,
                "points": item.points.iter().map(point_to_data).collect::<Vec<_>>(),
            })
        }).collect::<Vec<_>>(),
        "coordinate_system": result.transformer.coordinate_system.name(),
        "grid": result.transformer.grid.subdivisions.iter().map(|(axis, subdivision)| {
            (axis.clone(), json!({
                "minimum": subdivision.range.minimum,
                "maximum": subdivision.range.maximum,
                "steps": subdivision.steps,
            }))
        }).collect::<BTreeMap<_, _>>(),
    })
}

pub fn build_session_data(
    config: &ExplorerConfig,
    cases: &[ExplorationCase],
    results: &[(ExplorationCase, DiscreteAnalysisResult)],
    active_case_id: Option<String>,
    table_mode: &str,
    map_limits: &BTreeMap<String, (f64, f64, f64, f64)>,
    music: &BTreeMap<String, String>,
) -> Value {
    json!({
        "schema": SESSION_SCHEMA,
        "version": SESSION_VERSION,
        "view": {
            "config": config_to_data(config),
            "active_case_id": active_case_id,
            "table_mode": table_mode,
            "map_limits": map_limits.iter().map(|(key, value)| {
                (key.clone(), vec![value.0, value.1, value.2, value.3])
            }).collect::<BTreeMap<_, _>>(),
            "music": music,
        },
        "cases": cases.iter().map(case_to_data).collect::<Vec<_>>(),
        "results": results.iter().map(|(case, result)| {
            json!({
                "case_id": case.identifier,
                "case_label": case.label,
                "result": result_to_data(result),
            })
        }).collect::<Vec<_>>(),
    })
}

pub fn parse_session_data(data: &Value) -> Result<ImportedSession> {
    let object = data
        .as_object()
        .ok_or_else(|| anyhow::anyhow!("La raiz de la sesion debe ser un objeto JSON"))?;
    if object.get("schema").and_then(Value::as_str) != Some(SESSION_SCHEMA) {
        bail!("El archivo no es una sesion de fractal_sequences");
    }
    if object.get("version").and_then(Value::as_u64) != Some(SESSION_VERSION as u64) {
        bail!("Version de sesion no compatible");
    }
    let view = object
        .get("view")
        .ok_or_else(|| anyhow::anyhow!("Sesion incompleta"))?;
    let view_object = view
        .as_object()
        .ok_or_else(|| anyhow::anyhow!("Sesion incompleta"))?;
    let config = config_from_data(
        view_object
            .get("config")
            .ok_or_else(|| anyhow::anyhow!("Sesion incompleta"))?,
    )?;
    let cases_value = object
        .get("cases")
        .ok_or_else(|| anyhow::anyhow!("Sesion incompleta"))?;
    let case_items = cases_value
        .as_array()
        .ok_or_else(|| anyhow::anyhow!("Sesion incompleta"))?;
    let cases = case_items
        .iter()
        .map(case_from_data)
        .collect::<Result<Vec<_>>>()?;

    let mut map_limits = BTreeMap::new();
    if let Some(raw_limits) = view_object.get("map_limits").and_then(Value::as_object) {
        for (key, value) in raw_limits {
            if let Some(items) = value.as_array()
                && items.len() == 4
            {
                let parsed = (
                    items[0].as_f64().unwrap_or(0.0),
                    items[1].as_f64().unwrap_or(0.0),
                    items[2].as_f64().unwrap_or(0.0),
                    items[3].as_f64().unwrap_or(0.0),
                );
                map_limits.insert(key.clone(), parsed);
            }
        }
    }

    let table_mode = view_object
        .get("table_mode")
        .and_then(Value::as_str)
        .unwrap_or("todo")
        .to_string();
    let active_case_id = view_object
        .get("active_case_id")
        .and_then(Value::as_str)
        .map(ToString::to_string);
    let music = if let Some(music_object) = view_object.get("music").and_then(Value::as_object) {
        music_object
            .iter()
            .map(|(key, value)| (key.clone(), value.as_str().unwrap_or("").to_string()))
            .collect()
    } else {
        BTreeMap::new()
    };

    Ok(ImportedSession {
        config,
        cases,
        active_case_id,
        table_mode,
        map_limits,
        music,
    })
}

pub fn write_session(path: impl AsRef<Path>, data: &Value) -> Result<()> {
    std::fs::write(path, serde_json::to_string_pretty(data)?)?;
    Ok(())
}

pub fn read_session(path: impl AsRef<Path>) -> Result<ImportedSession> {
    let raw = std::fs::read_to_string(path)?;
    let data: Value = serde_json::from_str(&raw)?;
    parse_session_data(&data)
}

pub fn write_results_csv(
    path: impl AsRef<Path>,
    results: &[(ExplorationCase, DiscreteAnalysisResult)],
) -> Result<()> {
    let mut writer = csv::Writer::from_path(path)?;
    writer.write_record([
        "case_id",
        "case_label",
        "source",
        "row_type",
        "attractor_index",
        "term_index",
        "value_real",
        "value_imag",
        "coordinates",
        "indices",
        "discrete_coordinates",
        "cell_id",
        "inside_domain",
        "attractor_kind",
        "attractor_period",
    ])?;

    for (case, result) in results {
        for item in &result.terms {
            write_csv_row(&mut writer, case, "sequence", &item.point, None, Some(item.term.index), None, None)?;
        }
        for (attractor_index, attractor) in result.attractors.iter().enumerate() {
            for point in &attractor.points {
                write_csv_row(
                    &mut writer,
                    case,
                    "attractor",
                    point,
                    Some(attractor_index + 1),
                    None,
                    Some(format!("{:?}", attractor.attractor.kind).to_lowercase()),
                    Some(attractor.attractor.period),
                )?;
            }
        }
    }

    writer.flush()?;
    Ok(())
}

fn write_csv_row(
    writer: &mut csv::Writer<std::fs::File>,
    case: &ExplorationCase,
    row_type: &str,
    point: &DiscretePoint,
    attractor_index: Option<usize>,
    term_index: Option<usize>,
    attractor_kind: Option<String>,
    attractor_period: Option<usize>,
) -> Result<()> {
    let value = point.value;
    writer.write_record([
        case.identifier.clone(),
        case.label.clone(),
        case.source.clone(),
        row_type.to_string(),
        attractor_index.map_or_else(String::new, |value| value.to_string()),
        term_index.map_or_else(String::new, |value| value.to_string()),
        value.re.to_string(),
        value.im.to_string(),
        serde_json::to_string(&point.coordinates)?,
        serde_json::to_string(&point.indices)?,
        serde_json::to_string(&point.discrete_coordinates)?,
        point.cell_id.map_or_else(String::new, |value| value.to_string()),
        point.inside_domain.to_string(),
        attractor_kind.unwrap_or_default(),
        attractor_period.map_or_else(String::new, |value| value.to_string()),
    ])?;
    Ok(())
}

fn term_to_data(term: &SequenceTerm) -> Value {
    json!({
        "index": term.index,
        "value": number_to_data(term.value),
    })
}

fn attractor_to_data(attractor: &Attractor) -> Value {
    json!({
        "values": attractor.values.iter().map(|value| number_to_data(*value)).collect::<Vec<_>>(),
        "kind": attractor.kind,
        "period": attractor.period,
        "tolerance": attractor.tolerance,
        "start_index": attractor.start_index,
    })
}

fn point_to_data(point: &DiscretePoint) -> Value {
    json!({
        "value": number_to_data(point.value),
        "coordinates": point.coordinates,
        "indices": point.indices,
        "discrete_coordinates": point.discrete_coordinates,
        "cell_id": point.cell_id,
        "inside_domain": point.inside_domain,
    })
}

fn number_to_data(value: crate::models::Number) -> Value {
    json!({
        "real": value.re,
        "imag": value.im,
    })
}
